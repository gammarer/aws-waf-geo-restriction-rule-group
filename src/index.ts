import * as waf from 'aws-cdk-lib/aws-wafv2';
import { Construct } from 'constructs';

export interface WafGeoRestrictRuleGroupProps {
  readonly name?: string;
  readonly scope: Scope;
  readonly allowCountries: string[];
  readonly ipRateLimiting?: IpRateLimitingProperty;
  //readonly rateLimitCount?: number;
  //whitelist
}

export interface IpRateLimitingProperty {
  readonly enable: boolean;
  readonly count: number;
}

export enum Scope {
  GLOBAL = 'Global',
  REGIONAL = 'Regional',
}

export class WafGeoRestrictRuleGroup extends waf.CfnRuleGroup {
  constructor(scope: Construct, id: string, props: WafGeoRestrictRuleGroupProps) {
    super(scope, id, {
      name: props.name,
      description: 'geo restrict rule group',
      scope: ((): string => {
        switch (props.scope) {
          case Scope.GLOBAL:
            return 'CLOUDFRONT';
          case Scope.REGIONAL:
            return 'REGIONAL';
        }
      })(),
      capacity: 10,
      rules: ((): Array<waf.CfnRuleGroup.RuleProperty> => {
        const defaults: Array<waf.CfnRuleGroup.RuleProperty> = [
          {
            priority: 0,
            name: 'allow-geo-rule',
            action: {
              allow: {},
            },
            visibilityConfig: {
              cloudWatchMetricsEnabled: true,
              sampledRequestsEnabled: true,
              metricName: 'AllowGeoRule',
            },
            statement: {
              geoMatchStatement: {
                countryCodes: props.allowCountries,
              },
            },
          },
          {
            priority: 5,
            name: 'deny-geo-rule',
            action: {
              block: {
                CustomResponse: {
                  CustomResponseBodyKey: 'geo-restrict',
                  ResponseCode: 403,
                },
              },
            },
            visibilityConfig: {
              cloudWatchMetricsEnabled: true,
              sampledRequestsEnabled: true,
              metricName: 'DenyGeoRule',
            },
            statement: {
              notStatement: {
                statement: {
                  geoMatchStatement: {
                    countryCodes: props.allowCountries,
                  },
                },
              },
            },
          },
        ];
        if (props.ipRateLimiting) {
          if (props.ipRateLimiting.enable) {
            if (props.ipRateLimiting.count < 100) {
              throw new Error('IP rate limiting count value needs to be above 100');
            } else {
              const rule: waf.CfnRuleGroup.RuleProperty = {
                priority: 2,
                name: 'ip-rate-limiting-rule',
                action: {
                  block: {
                    CustomResponse: {
                      CustomResponseBodyKey: 'ip-restrict',
                      ResponseCode: 403,
                    },
                  },
                },
                visibilityConfig: {
                  cloudWatchMetricsEnabled: true,
                  sampledRequestsEnabled: true,
                  metricName: 'DenyIpLimitingRule',
                },
                statement: {
                  rateBasedStatement: {
                    aggregateKeyType: 'IP',
                    limit: props?.ipRateLimiting.count,
                  },
                },
              };
              defaults.push(rule);
            }
          }
        }
        return defaults;
      })(),
      customResponseBodies: {
        ['geo-restrict']: {
          contentType: 'TEXT_PLAIN',
          content: 'Sorry, You Are Not Allowed to Access This Service.',
        },
      },
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        sampledRequestsEnabled: true,
        metricName: 'GeoRestrictRule',
      },
    });
  }
}