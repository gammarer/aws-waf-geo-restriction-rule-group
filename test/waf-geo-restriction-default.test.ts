import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as waf from 'aws-cdk-lib/aws-wafv2';
import { Scope, WafGeoRestrictRuleGroup } from '../src';

describe('Web ACL Rule Group default testing', () => {

  describe('Web Acl Rule Group scope=global testing', () => {

    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    const ruleGroup = new WafGeoRestrictRuleGroup(stack, 'WafGeoRestrictRuleGroup', {
      scope: Scope.GLOBAL,
      countryCodes: ['JP'],
    });

    it('Is Waf RuleGroup', () => {
      expect(ruleGroup).toBeInstanceOf(waf.CfnRuleGroup);
    });

    const template = Template.fromStack(stack);

    it('Should have WAF Rule Group', () => {
      template.hasResourceProperties('AWS::WAFv2::RuleGroup', Match.objectEquals({
        Description: 'geo restrict rule group',
        Scope: 'CLOUDFRONT',
        Capacity: 10,
        CustomResponseBodies: {
          'geo-restrict': {
            Content: 'Sorry, You Are Not Allowed to Access This Service.',
            ContentType: 'TEXT_PLAIN',
          },
        },
        Rules: [
          {
            Priority: 10,
            Name: 'geo-restrict-rule',
            Action: {
              Block: {
                CustomResponse: {
                  CustomResponseBodyKey: 'geo-restrict',
                  ResponseCode: 403,
                },
              },
            },
            VisibilityConfig: {
              CloudWatchMetricsEnabled: true,
              MetricName: 'WafGeoRestrictRule',
              SampledRequestsEnabled: true,
            },
            Statement: {
              GeoMatchStatement: {
                CountryCodes: ['JP'],
              },
            },
          },
        ],
        VisibilityConfig: {
          CloudWatchMetricsEnabled: true,
          MetricName: 'GeoRestrictRule',
          SampledRequestsEnabled: true,
        },
      }));
    });

    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('global');
    });

  });

  describe('Web Acl Rule Group scope=regional testing', () => {

    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    const ruleGroup = new WafGeoRestrictRuleGroup(stack, 'WafGeoRestrictRuleGroup', {
      name: 'example-waf-geo-restrict-rule-group',
      scope: Scope.REGIONAL,
      countryCodes: ['JP'],
    });

    it('Is Waf RuleGroup', () => {
      expect(ruleGroup).toBeInstanceOf(waf.CfnRuleGroup);
    });

    const template = Template.fromStack(stack);

    it('Should have WAF Rule Group', () => {
      template.hasResourceProperties('AWS::WAFv2::RuleGroup', Match.objectEquals({
        Name: 'example-waf-geo-restrict-rule-group',
        Description: 'geo restrict rule group',
        Scope: 'REGIONAL',
        Capacity: 10,
        CustomResponseBodies: {
          'geo-restrict': {
            Content: 'Sorry, You Are Not Allowed to Access This Service.',
            ContentType: 'TEXT_PLAIN',
          },
        },
        Rules: [
          {
            Priority: 10,
            Name: 'geo-restrict-rule',
            Action: {
              Block: {
                CustomResponse: {
                  CustomResponseBodyKey: 'geo-restrict',
                  ResponseCode: 403,
                },
              },
            },
            VisibilityConfig: {
              CloudWatchMetricsEnabled: true,
              MetricName: 'WafGeoRestrictRule',
              SampledRequestsEnabled: true,
            },
            Statement: {
              GeoMatchStatement: {
                CountryCodes: ['JP'],
              },
            },
          },
        ],
        VisibilityConfig: {
          CloudWatchMetricsEnabled: true,
          MetricName: 'GeoRestrictRule',
          SampledRequestsEnabled: true,
        },
      }));
    });

    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('regional');
    });

  });

});