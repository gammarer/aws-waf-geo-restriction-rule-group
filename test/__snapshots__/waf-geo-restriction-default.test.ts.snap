// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web ACL Rule Group default testing Web Acl Rule Group scope=global testing Should match snapshot: global 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "WafGeoRestrictRuleGroup": Object {
      "Properties": Object {
        "Capacity": 10,
        "CustomResponseBodies": Object {
          "geo-restrict": Object {
            "Content": "Sorry, You Are Not Allowed to Access This Service.",
            "ContentType": "TEXT_PLAIN",
          },
        },
        "Description": "geo restrict rule group",
        "Rules": Array [
          Object {
            "Action": Object {
              "Block": Object {
                "CustomResponse": Object {
                  "CustomResponseBodyKey": "geo-restrict",
                  "ResponseCode": 403,
                },
              },
            },
            "Name": "geo-restrict-rule",
            "Priority": 10,
            "Statement": Object {
              "GeoMatchStatement": Object {
                "CountryCodes": Array [
                  "JP",
                ],
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "WafGeoRestrictRule",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "CLOUDFRONT",
        "VisibilityConfig": Object {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "GeoRestrictRule",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::RuleGroup",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Web ACL Rule Group default testing Web Acl Rule Group scope=regional testing Should match snapshot: regional 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "WafGeoRestrictRuleGroup": Object {
      "Properties": Object {
        "Capacity": 10,
        "CustomResponseBodies": Object {
          "geo-restrict": Object {
            "Content": "Sorry, You Are Not Allowed to Access This Service.",
            "ContentType": "TEXT_PLAIN",
          },
        },
        "Description": "geo restrict rule group",
        "Name": "example-waf-geo-restrict-rule-group",
        "Rules": Array [
          Object {
            "Action": Object {
              "Block": Object {
                "CustomResponse": Object {
                  "CustomResponseBodyKey": "geo-restrict",
                  "ResponseCode": 403,
                },
              },
            },
            "Name": "geo-restrict-rule",
            "Priority": 10,
            "Statement": Object {
              "GeoMatchStatement": Object {
                "CountryCodes": Array [
                  "JP",
                ],
              },
            },
            "VisibilityConfig": Object {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "WafGeoRestrictRule",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "REGIONAL",
        "VisibilityConfig": Object {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "GeoRestrictRule",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::RuleGroup",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
